backtrace-library/src/main/java/backtraceio/library/breadcrumbs/BacktraceBreadcrumbs.java

```diff
@@ -0,0 +1,246 @@
+package backtraceio.library.breadcrumbs;
+
+import android.app.Activity;
+import android.app.Application;
+import android.content.Context;
+import android.support.annotation.NonNull;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import backtraceio.library.enums.BacktraceBreadcrumbLevel;
+import backtraceio.library.enums.BacktraceBreadcrumbType;
+import backtraceio.library.interfaces.Breadcrumbs;
+
+public class BacktraceBreadcrumbs implements Breadcrumbs {
+
+    /**
+     * Which breadcrumb types are enabled?
+     */
+    private Set<BacktraceBreadcrumbType> enabledBreadcrumbTypes;
+
+    /**
+     * The Backtrace BroadcastReciever instance
+     */
+    private BacktraceBroadcastReceiver backtraceBroadcastReceiver;
+
+    /**
+     * The Backtrace ComponentCallbacks2 listener
+     */
+    private BacktraceComponentListener backtraceComponentListener;
+
+    /**
+     * The Backtrace ActivityLifecycleCallbacks listener
+     */
+    private BacktraceActivityLifecycleListener backtraceActivityLifecycleListener;
+
+    /**
+     * The Backtrace Breadcrumbs logger instance
+     */
+    private BacktraceBreadcrumbsLogger backtraceBreadcrumbsLogger;
+
+    final private static String breadcrumbLogDirectory = "breadcrumbs";
+
+    private Context context;
+
+    private static final int DEFAULT_MAX_LOG_SIZE_BYTES = 64000;
+
+    public BacktraceBreadcrumbs(Context context, int maxBreadcrumbLogSizeBytes) throws IOException, NoSuchMethodException {
+        // Create the breadcrumbs subdirectory for storing the breadcrumb logs
+        this.context = context;
+        String fullBreadcrumbLogDirectory = getBreadcrumbLogDirectory(context);
+        backtraceBreadcrumbsLogger = new BacktraceBreadcrumbsLogger(fullBreadcrumbLogDirectory, maxBreadcrumbLogSizeBytes);
+    }
+
+    public BacktraceBreadcrumbs(Context context) throws IOException, NoSuchMethodException {
+        this(context, DEFAULT_MAX_LOG_SIZE_BYTES);
+    }
+
+    private void registerAutomaticBreadcrumbReceivers(Set<BacktraceBreadcrumbType> breadcrumbTypesToEnable) {
+        if (isBreadcrumbsEnabled()) {
+            unregisterAutomaticBreadcrumbReceivers();
+        }
+
+        backtraceBroadcastReceiver = new BacktraceBroadcastReceiver(this);
+        context.registerReceiver(backtraceBroadcastReceiver,
```

Comment:
we can run `setEnabledBreadcrumbTypes` mutliple times - does it mean that we can receive duplicated breadcrumbs?

Reply:
We prevent this by requiring `isNewInstance` check, if we have a new `BacktraceBreadcrumbs` instance, we should register the receivers. If it is not a new `BacktraceBreadcrumbs` instance, we should call `unregisterAutomaticBreadcrumbReceivers` first. Thank you for checking this! I think it's more clear now too after refactoring according to your above suggestions!